singularity-eos/eos/eos_carnahan_starling.hpp

```diff
@@ -0,0 +1,258 @@
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_CARNAHAN_STARLING_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_CARNAHAN_STARLING_HPP_
+
+// stdlib
+#include <cmath>
+#include <cstdio>
+#include <string>
+
+// Ports-of-call
+#include <ports-of-call/portability.hpp>
+
+// Base stuff
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/base/root-finding-1d/root_finding.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+class CarnahanStarling : public EosBase<CarnahanStarling> {
+ public:
+  CarnahanStarling() = default;
+  PORTABLE_INLINE_FUNCTION CarnahanStarling(Real gm1, Real Cv, Real bb, Real qq)
+      : _Cv(Cv), _gm1(gm1), _bb(bb), _qq(qq), _T0(ROOM_TEMPERATURE),
+        _P0(ATMOSPHERIC_PRESSURE), _qp(0.0),
+        _rho0(DensityFromPressureTemperature(_P0, _T0)), _vol0(robust::ratio(1.0, _rho0)),
+        _sie0(Cv * _T0 + qq),
+        _bmod0(_rho0 * Cv * _T0 *
+               (PartialRhoZedFromDensity(_rho0) +
+                ZedFromDensity(_rho0) * ZedFromDensity(_rho0) * gm1)),
+        _dpde0(_rho0 * ZedFromDensity(_rho0) * gm1) {
+    checkParams();
+  }
+  PORTABLE_INLINE_FUNCTION CarnahanStarling(Real gm1, Real Cv, Real bb, Real qq, Real qp,
+                                            Real T0, Real P0)
+      : _Cv(Cv), _gm1(gm1), _bb(bb), _qq(qq), _T0(T0), _P0(P0), _qp(qp),
+        _rho0(DensityFromPressureTemperature(P0, T0)), _vol0(robust::ratio(1.0, _rho0)),
+        _sie0(Cv * T0 + qq),
+        _bmod0(_rho0 * Cv * T0 *
+               (PartialRhoZedFromDensity(_rho0) +
+                ZedFromDensity(_rho0) * ZedFromDensity(_rho0) * gm1)),
+        _dpde0(_rho0 * ZedFromDensity(_rho0) * gm1) {
+    checkParams();
+  }
+  CarnahanStarling GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return std::max(robust::SMALL(), (sie - _qq) / _Cv);
+  }
+  PORTABLE_INLINE_FUNCTION void checkParams() const {
+    PORTABLE_ALWAYS_REQUIRE(_Cv >= 0, "Heat capacity must be positive");
+    PORTABLE_ALWAYS_REQUIRE(_gm1 >= 0, "Gruneisen parameter must be positive");
+    PORTABLE_ALWAYS_REQUIRE(_bb >= 0, "Covolume must be positive");
+  }
+  PORTABLE_INLINE_FUNCTION Real ZedFromDensity(const Real rho,
+                                               Real *lambda = nullptr) const {
+    const Real eta = _bb * rho;
+    const Real eta2 = eta * eta;
+    const Real zed = robust::ratio(1.0 + eta + eta2 - eta2 * eta,
```

Comment:
I think this one should be `1.0 + robust::ratio(x*(4.0 - 2.0*x), singularity::math_utils::pow<3>(1.0 - x));`

Reply:
alternatively, I think `robust::ratio(1.0 + eta*(1.0 + eta*(1.0 - eta)), singularity::math_utils::pow<3>(1.0 - eta));` should be just as efficient assuming we get fma's, this is not true for quartic, Horner's method for evaluating the polynomial appears to result in 1 more instruction.

Reply:
Resolved